02-KernelModules_CharacterDevice/chrdev.c

```
@@ -0,0 +1,234 @@
+	ssize_t retval = 0;
+
+	int remain = buffer_size - (int) (*offset);
+	if (remain == 0) {
+		return retval;
+	}
+
+	if (len > remain) {
+		len = remain;
+	}
+
+	if (raw_copy_to_user(buffer, data_buff + *offset, len)) {
+		retval = -EFAULT;
+		return retval;
+	}
+	*offset += len;
+	retval = len;
+
+	return retval;
+}
+
+ssize_t dev_write(struct file *filep, const char *buffer, size_t len, loff_t *offset){
+	int error = 0;
+	int remain = buffer_size - (int) (*offset);
+	
+	used_memory += len;
+	
+	if(0 == remain) return 0;
+	if(used_memory > remain) return -EIO;
+	
+	error = raw_copy_from_user(data_buff + *offset, buffer, len);
+	if(error){
+		printk(KERN_INFO "chrdev: failed to receive %zu characters from the user\n", len);
+		return -EFAULT;
+	}
+	else{
+		printk(KERN_INFO "chrdev: received %zu characters from the user\n", len);
+		offset += len;
+		return len;
+	}
+}
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
+
+static int __init chrdev_init(void){
+	int err;
+	
+	printk(KERN_INFO "chrdev loading");
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	if (major < 0) {
+		pr_err("register_chrdev failed: %d\n", major);
+		return major;
+	}
+	pr_info("chrdev: register_chrdev ok, major = %d\n", major);
+	
+	pclass = class_create(THIS_MODULE, CLASS_NAME);
+	if (IS_ERR(pclass)) {
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: class_create failed\n");
+		return PTR_ERR(pclass);
+	}
+	pr_info("chrdev: device class created successfully\n");
+
+    pdev = device_create(pclass, NULL, MKDEV(major, 0), NULL, DEVICE_NAME);
+	if (IS_ERR(pdev)) {
+		class_destroy(pclass);
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: device_create failed\n");
+		return PTR_ERR(pdev);
+	}
+	printk(KERN_INFO "chrdev: Device initialized succesfully\n");
+
+	buffer_create();
```

The functions buffer_crerate() and create_proc() can return errors, but these errors are ignored here.